### PR TITLE
Support lambda expressions and method groups as `var` initializers

### PR DESCRIPTION
**Pull request by cston**
```csharp
var d1 = Program.Main;
var d2 = () => { };
```

Proposal: [lambda-improvements.md](https://github.com/dotnet/csharplang/blob/main/proposals/lambda-improvements.md)
Test plan: https://github.com/dotnet/roslyn/issues/52192

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs [hunk 1]
```diff
@@ -1991,29 +1996,301 @@ static void Main()
...
+                // (5,28): error CS0446: Foreach cannot operate on a 'method group'. Did you intend to invoke the 'method group'?
+                //         foreach (var d1 in Main) { }
+                Diagnostic(ErrorCode.ERR_AnonMethGrpInForEach, "Main").WithArguments("method group").WithLocation(5, 28),
+                // (6,28): error CS0446: Foreach cannot operate on a 'lambda expression'. Did you intend to invoke the 'lambda expression'?
+                //         foreach (var d2 in () => { }) { }
+                Diagnostic(ErrorCode.ERR_AnonMethGrpInForEach, "() => { }").WithArguments("lambda expression").WithLocation(6, 28),
+                // (7,28): error CS0446: Foreach cannot operate on a 'anonymous method'. Did you intend to invoke the 'anonymous method'?
+                //         foreach (var d3 in delegate () { }) { }
+                Diagnostic(ErrorCode.ERR_AnonMethGrpInForEach, "delegate () { }").WithArguments("anonymous method").WithLocation(7, 28));
+        }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:2191]: nit: consider another similar negative test with deconstruction declaration `(var x, var y) = ...`

Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:2191]: Also, something like `(Action a, var b) = (null, () => {});` (I assume should work per spec, but may not work in current implementation)

Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:2191]: Please check `GetTypeInfo` and `GetDeclaredSymbol` in one of the positive tests

Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:2191]: Consider including a test with a non-simple signature (error)

File: src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs [hunk 2]
```diff
@@ -11934,24 +11934,15 @@ public static void Main()
     }
     static void M() {}
 }").VerifyDiagnostics(
-                // (6,13): error CS0815: Cannot assign method group to an implicitly-typed variable
-                //         var m = Main; // CS0815
-                Diagnostic(ErrorCode.ERR_ImplicitlyTypedVariableAssignedBadValue, "m = Main").WithArguments("method group"),
-                // (7,13): error CS0815: Cannot assign lambda expression to an implicitly-typed variable
+                // (7,17): error CS8917: The delegate type could not be inferred.
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs:11937]: Comment in `var m = Main; // CS0815` is out-of-date

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs (hunk 1 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:2191]: Added tests. I'll investigate the failures with the tuple cases.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 2). How do you want to track the deconstruction issue?

**Comment by jcouv** [comment]
@333fred @dotnet/roslyn-compiler for a second review. Thanks

**Comment by cston** [comment]
> How do you want to track the deconstruction issue?

Logged https://github.com/dotnet/roslyn/issues/55139.

**Comment by 333fred** [comment]
Looking.

**Comment by 333fred** [comment]
>         public void IAnonymousFunctionExpression_UnboundLambdaAsVar_HasValidLambdaExpressionTree()

Nit: consider renaming to BoundLambda

---
Refers to: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IAnonymousFunctionExpression.cs:190 in a884c74. [](commit_id = a884c7484a073d4d6268e43b1256f4fe6de02804, deletion_comment = False)

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IAnonymousFunctionExpression.cs [hunk 3]
```diff
@@ -409,15 +410,15 @@ static void F()
             var lambdaSyntax = (LambdaExpressionSyntax)variableDeclaration.Declaration.Variables.Single().Initializer.Value;
 
             var variableDeclarationGroupOperation = (IVariableDeclarationGroupOperation)semanticModel.GetOperation(variableDeclaration);
-            var variableTreeLambdaOperation = (IAnonymousFunctionOperation)variableDeclarationGroupOperation.Declarations.Single().Declarators.Single().Initializer.Value;
+            var variableTreeLambdaOperation = ((IDelegateCreationOperation)variableDeclarationGroupOperation.Declarations.Single().Declarators.Single().Initializer.Value).Target;
```
Comment (333fred) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IAnonymousFunctionExpression.cs:413]: >IDelegateCreationOperation [](http://example.com/codeflow?start=48&length=26)

I don't think this is keeping the test correct. Please break the test in a different fashion, such as removing the `F()` definition so the body is still unbound.

**Comment by 333fred** [comment]
>         public void InvalidLambdaBinding_UnboundLambda()

No longer correctly named, and probably not the correct location for this either, since it's no longer invalid. This test can probably just be removed.

---
Refers to: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_InvalidExpression.cs:333 in a884c74. [](commit_id = a884c7484a073d4d6268e43b1256f4fe6de02804, deletion_comment = False)

**Comment by 333fred** [comment]
>         public void InvalidLambdaBinding_LambdaExpression()

Same comment.

---
Refers to: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_InvalidExpression.cs:375 in a884c74. [](commit_id = a884c7484a073d4d6268e43b1256f4fe6de02804, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs [hunk 4]
```diff
@@ -1991,29 +1996,438 @@ static void Main()
...
+                // (7,17): error CS8172: Cannot initialize a by-reference variable with a value
+                //         ref var d3 = delegate () { };
+                Diagnostic(ErrorCode.ERR_InitializeByReferenceVariableWithValue, "d3 = delegate () { }").WithLocation(7, 17),
+                // (7,22): error CS1510: A ref or out value must be an assignable variable
+                //         ref var d3 = delegate () { };
+                Diagnostic(ErrorCode.ERR_RefLvalueExpected, "delegate () { }").WithLocation(7, 22));
+        }
+
+        [Fact]
+        public void ImplicitlyTypedVariables_04()
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:2142]: >ImplicitlyTypedVariables_04 [](http://example.com/codeflow?start=20&length=27)

Would it be worth making a version of this test where `Action` inherits from `IDisposable`?

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs [hunk 5]
```diff
@@ -1991,29 +1996,438 @@ static void Main()
...
+        public void ImplicitlyTypedVariables_12()
+        {
+            var source =
+@"class Program
+{
+    static void F(ref int i) { }
+    static void Main()
+    {
+        var d1 = F;
+        var d2 = (ref int x) => x;
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:2372]: Consider expanding these negative tests with ref returns as well.

**Review by 333fred — APPROVED** [review]
I'm not going to block on any of my test suggestions, but please make sure they're in the next PR.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IAnonymousFunctionExpression.cs (hunk 3 above)
Comment (cston) [reply src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IAnonymousFunctionExpression.cs:413]: Will add to the next PR, thanks.

**Comment by cston** [comment]
>         public void IAnonymousFunctionExpression_UnboundLambdaAsVar_HasValidLambdaExpressionTree()

Will add to next PR.

---
In reply to: [887712078](https://github.com/dotnet/roslyn/pull/55132#issuecomment-887712078) [](http://example.com/codeflow?ancestors=887712078,887712078)

---
Refers to: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IAnonymousFunctionExpression.cs:190 in a884c74. [](commit_id = a884c7484a073d4d6268e43b1256f4fe6de02804, deletion_comment = False)

**Comment by cston** [comment]
>         public void InvalidLambdaBinding_UnboundLambda()

Will leave these tests here, to catch any regressions, but I'll rename the tests.

---
In reply to: [887713598](https://github.com/dotnet/roslyn/pull/55132#issuecomment-887713598) [](http://example.com/codeflow?ancestors=887713598,887713598)

---
Refers to: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_InvalidExpression.cs:333 in a884c74. [](commit_id = a884c7484a073d4d6268e43b1256f4fe6de02804, deletion_comment = False)

**Comment by davidfowl** [comment]
Tell me when this is merged 😄

**Comment by oskardudycz** [comment]
@cston, a small correction. The link to the proposal should be https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/lambda-improvements.md.
This looks neat! 😄

**Pull request merged**